File: Codes/Matlab/Chapters/Chapter8/Circulation_ss.m


% =============================================================================
% Created on Tue Oct 26 11:34:10 2021
% Circulation model: Algebraic models demonstrating sampling methods
% @author: cogan
% =============================================================================

clear
close all
set(groot,'DefaultAxesTickLabelInterpreter','latex');
set(groot,'DefaultLegendInterpreter','latex');
options = optimoptions(@fsolve, 'MaxFunctionEvaluations', 10000, 'Display', 'off');
% =============================================================================
% We can find at least one root
% =============================================================================
params=[5, 5.5, 1.50, 11,2,80, 5,4, .5,.12];
solution1 = fsolve(@(Y)rhs(Y,params), [10,.11,.1],options)
% =============================================================================
% Normal distribution
% =============================================================================
n_power=10;
Num_samples=2^n_power;
total_bacteria=zeros(1,Num_samples);


parameters_normal=lhsnorm(params,diag(.01*params),Num_samples);
% =============================================================================
% Uniform distribution
% =============================================================================
% =============================================================================
% Note that lhsdesign is defined on the unit interval
% =============================================================================
parameters_uniform_unscaled = lhsdesign(Num_samples,length(params));
% =============================================================================
% Scale the samples into the correct parameter scale
% =============================================================================
l_bounds = .95*params;
u_bounds = 1.05*params;
parameters_uniform=zeros(size(parameters_uniform_unscaled));
for i=1:Num_samples
    parameters_uniform(i,:)=(u_bounds-l_bounds).*parameters_uniform_unscaled(i,:)+l_bounds;
end
% =============================================================================
% Sobol distribution: Sobol sequences are low discrepancy sequence
% sobolset is a Matlab native function with many variants -- this is a
% common one
% =============================================================================
p = sobolset(length(params),'Skip',1e3,'Leap',1e2);
p = scramble(p,'MatousekAffineOwen');
parameters_sobol_unscaled=net(p,Num_samples);
% =============================================================================
% Scale the samples into the correct parameter scale
% =============================================================================
l_bounds = .95*params;
u_bounds = 1.05*params;
parameters_sobol=zeros(size(parameters_sobol_unscaled));
for i=1:Num_samples
    parameters_sobol(i,:)=(u_bounds-l_bounds).*parameters_sobol_unscaled(i,:)+l_bounds;
end
FP_uniform=zeros(3,Num_samples);
FP_normal=zeros(3,Num_samples);
FP_sobol=zeros(3,Num_samples);


for i=1:Num_samples
    [FP_uniform(:,i),flag1]=fsolve(@(Y)rhs(Y,parameters_uniform(i,:)), [60,2.5,.1],options);
    [FP_normal(:,i),flag2]=fsolve(@(Y)rhs(Y,parameters_normal(i,:)), [60,2.5,.1],options);
    [FP_sobol(:,i),flag3]=fsolve(@(Y)rhs(Y,parameters_sobol(i,:)), [60,2.5,.1],options);
end
figure()
histogram(FP_uniform(1,:),20);
hold on
histogram(FP_normal(1,:),20);
histogram(FP_sobol(1,:),20);
xlabel('$QoI=Q$', fontsize=16)
ylabel('Frequency', fontsize=16)
legend('Uniform', 'Normal', 'Sobol')

figure()
histogram(FP_uniform(2,:),20);
hold on
histogram(FP_normal(2,:),20);
histogram(FP_sobol(2,:),20);
xlabel('$QoI=Q$', fontsize=16)
ylabel('Frequency', fontsize=16)
legend('Uniform', 'Normal', 'Sobol')

figure()
histogram(FP_uniform(3,:),20);
hold on
histogram(FP_normal(3,:),20);
histogram(FP_sobol(3,:),20);
xlabel('$QoI=Q$', fontsize=16)
ylabel('Frequency', fontsize=16)
legend('Uniform', 'Normal', 'Sobol')

% =============================================================================
% Define the RHS of ODEs
% =============================================================================


function  f1 = rhs_Q(Y,params)
    Q=Y(1);
    Pa=Y(2);
    Pv=Y(3);
    Qvessel=params(1);
    R=params(2);
    gamma=params(3);
    Vvessel=params(4);
    V0=params(5);
    F=params(6);
    Vmax=params(7);
    Vmin=params(8);
    Cd=params(9);
    Cs=params(10);
    f1 = (Pa-Pv)/R*(1+gamma*(Pa+Pv)+gamma^2/3*(Pa^2+Pa*Pv+Pv^2))-Qvessel;
end

function  f2 = rhs_Pa(Y,params)
    Q=Y(1);
    Pa=Y(2);
    Pv=Y(3);
    Qvessel=params(1);
    R=params(2);
    gamma=params(3);
    Vvessel=params(4);
    V0=params(5);
    F=params(6);
    Vmax=params(7);
    Vmin=params(8);
    Cd=params(9);
    Cs=params(10);
    f2 = V0*(1+gamma/2*(Pa+Pv)+gamma^2/6*(Pa-Pv)^2)-Vvessel;
end

function  f3 = rhs_Pv(Y,params)
    Q=Y(1);
    Pa=Y(2);
    Pv=Y(3);
    Qvessel=params(1);
    R=params(2);
    gamma=params(3);
    Vvessel=params(4);
    V0=params(5);
    F=params(6);
    Vmax=params(7);
    Vmin=params(8);
    Cd=params(9);
    Cs=params(10);
    f3 = F*(Vmax-Vmin+Cd*Pv-Cs*Pa)-Q;
end

function  g = rhs(Y,params)
    Q=Y(1);
    Pa=Y(2);
    Pv=Y(3);
    Qvessel=params(1);
    R=params(2);
    gamma=params(3);
    Vvessel=params(4);
    V0=params(5);
    F=params(6);
    Vmax=params(7);
    Vmin=params(8);
    Cd=params(9);
    Cs=params(10);
    g = [rhs_Q(Y,params),rhs_Pa(Y,params),rhs_Pv(Y,params)];
end

   

File: Codes/Python/Chapters/Chapter8/Circulation_ss.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Oct 26 11:34:10 2021
Circulation model: Algebraic 
@author: cogan
"""



from IPython import get_ipython
get_ipython().magic('reset -sf')
import numpy as np
import matplotlib.pyplot as pyplot
from scipy.stats import qmc # for LHS
from scipy.optimize import fsolve
pyplot.close('all')
pyplot.style.use('tableau-colorblind10')  #Fixes colorscheme to be accessible


#Define the RHS of the equations
def f1(Y,params):
    Q,Pa, Pv=Y
    Qvessel, R, gamma, Vvessel,V0, F, Vmax, Vmin, Cd, Cs= params
    return (Pa-Pv)/R*(1+gamma*(Pa+Pv)+gamma**2/3*(Pa**2+Pa*Pv+Pv**2))-Qvessel
def f2(Y,params):
    Q,Pa, Pv=Y
    Qvessel, R, gamma, Vvessel,V0, F, Vmax, Vmin, Cd, Cs= params
    return V0*(1+gamma/2*(Pa+Pv)+gamma**2/6*(Pa-Pv)**2)-Vvessel
def f3(Y,params):
    Q,Pa, Pv=Y
    Qvessel, R, gamma, Vvessel,V0, F, Vmax, Vmin, Cd, Cs= params
    return F*(Vmax-Vmin+Cd*Pv-Cs*Pa)-Q
def rhs(Y,params):
    Q,Pa, Pv=Y
    Qvessel, R, gamma, Vvessel,V0, F, Vmax, Vmin, Cd, Cs= params
    return [f1(Y,params),f2(Y,params),f3(Y,params)]

params=[5, 5.5, 1.50, 11,2,80, 5,4, .5,.12]
solution1 = fsolve(rhs, (10,.11,.1),args=(params,) )
print(solution1)

# =============================================================================
# Normal distribution
# =============================================================================
n_power=10
Num_samples=2**n_power
total_bacteria=np.zeros(Num_samples)

import pyDOE as doe
from scipy.stats.distributions import norm
parameters_normal=doe.lhs(10,samples=Num_samples)

for i in np.arange(10):
     parameters_normal[:,i] = norm(loc=params[i], scale=.04*params[i]).ppf(parameters_normal[:, i]) #scaled parameters
#Explore how the variance changes things

# =============================================================================
# Uniform distribution
# =============================================================================
sampler = qmc.LatinHypercube(d=10) #Define the sampling method and parameter dimension
parameters_uniform = sampler.random(n=Num_samples)      #number of samples to take
 #Scale the samples into the correct parameter scale
l_bounds = np.multiply(params,.95)
u_bounds = np.multiply(params,1.05)
parameters_uniform=qmc.scale(parameters_uniform, l_bounds, u_bounds)
# 
# =============================================================================
# Sobol distribution: Scrambled Sobol Sequences.
# =============================================================================
sampler = qmc.Sobol(d=10,scramble=True,) #Define the sampling method and parameter dimension
parameters_sobol = sampler.random_base2(m=n_power)   #number of samples to take
 #Scale the samples into the correct parameter scale
l_bounds = np.multiply(params,.95)
u_bounds = np.multiply(params,1.05)
parameters_sobol=qmc.scale(parameters_sobol, l_bounds, u_bounds)

FP_uniform=np.zeros([3,Num_samples])
FP_normal=np.zeros([3,Num_samples])
FP_sobol=np.zeros([3,Num_samples])


for i in np.arange(Num_samples):
    FP_uniform[:,i]=fsolve(rhs, (10,.11,.1),args=(parameters_uniform[i,:],) )
    FP_normal[:,i]=fsolve(rhs, (10,.11,.1),args=(parameters_normal[i,:],) )
    FP_sobol[:,i]=fsolve(rhs, (10,.11,.1),args=(parameters_sobol[i,:],) )


pyplot.hist(FP_uniform[0,:],20, alpha=.5)
pyplot.hist(FP_normal[0,:],20, alpha=.5)
pyplot.hist(FP_sobol[0,:],20, alpha=.5)
pyplot.axis([0, 175,0, 300])
pyplot.xlabel('$QoI=Q$', fontsize=16)
pyplot.ylabel('Frequency', fontsize=16)
pyplot.legend(['Uniform', 'Normal', 'Sobol'])
pyplot.savefig('Q.png')

pyplot.figure()
pyplot.hist(FP_uniform[1,:],20, alpha=.5)
pyplot.hist(FP_normal[1,:],20, alpha=.5)
pyplot.hist(FP_sobol[1,:],20, alpha=.5)
pyplot.axis('tight')
pyplot.xlabel('$QoI=P_a$', fontsize=16)
pyplot.ylabel('Frequency', fontsize=16)
pyplot.legend(['Uniform', 'Normal', 'Sobol'])
pyplot.savefig('P_a.png')

pyplot.figure()

pyplot.hist(FP_uniform[2,:],20, alpha=.5)
pyplot.hist(FP_normal[2,:],20, alpha=.5)
pyplot.hist(FP_sobol[2,:],20, alpha=.5)
pyplot.axis('tight')
pyplot.xlabel('$QoI=P_v$', fontsize=16)
pyplot.ylabel('Frequency', fontsize=16)
pyplot.legend(['Uniform', 'Normal', 'Sobol'])
pyplot.savefig('P_v.png')


   